File: clients/go/model/game.go
package model

import (
	"io"

	mStream "../stream"
)

//Game -- global objects for game
type Game struct {
	CurrentTick int32
	Properties  *Properties
	Level       *Level
	Players     []*Player
	Units       []*Unit
	Bullets     []*Bullet
	Mines       []*Mine
	LootBoxes   []*LootBox
}

//NewGame -- return link to new Game
func NewGame(currentTick int32, properties *Properties, level *Level, players []*Player,
	units []*Unit, bullets []*Bullet, mines []*Mine, lootBoxes []*LootBox) *Game {
	return &Game{
		CurrentTick: currentTick,
		Properties:  properties,
		Level:       level,
		Players:     players,
		Units:       units,
		Bullets:     bullets,
		Mines:       mines,
		LootBoxes:   lootBoxes,
	}
}

//ReadGame -- read from net Game from LocalRunner
func ReadGame(reader io.Reader) *Game {
	result := &Game{
		CurrentTick: mStream.ReadInt32(reader),
		Properties:  ReadProperties(reader),
		Level:       ReadLevel(reader),
		Players:     make([]*Player, mStream.ReadInt32(reader)),
	}
	for i := range result.Players {
		result.Players[i] = ReadPlayer(reader)
	}
	result.Units = make([]*Unit, mStream.ReadInt32(reader))
	for i := range result.Units {
		result.Units[i] = ReadUnit(reader)
	}
	result.Bullets = make([]*Bullet, mStream.ReadInt32(reader))
	for i := range result.Bullets {
		result.Bullets[i] = ReadBullet(reader)
	}
	result.Mines = make([]*Mine, mStream.ReadInt32(reader))
	for i := range result.Mines {
		result.Mines[i] = ReadMine(reader)
	}
	result.LootBoxes = make([]*LootBox, mStream.ReadInt32(reader))
	for i := range result.LootBoxes {
		result.LootBoxes[i] = ReadLootBox(reader)
	}
	return result
}
func (value Game) Write(writer io.Writer) {
	mStream.WriteInt32(writer, value.CurrentTick)
	value.Properties.Write(writer)
	value.Level.Write(writer)
	mStream.WriteInt32(writer, int32(len(value.Players)))
	for _, PlayersElement := range value.Players {
		PlayersElement.Write(writer)
	}
	mStream.WriteInt32(writer, int32(len(value.Units)))
	for _, UnitsElement := range value.Units {
		UnitsElement.Write(writer)
	}
	mStream.WriteInt32(writer, int32(len(value.Bullets)))
	for _, BulletsElement := range value.Bullets {
		BulletsElement.Write(writer)
	}
	mStream.WriteInt32(writer, int32(len(value.Mines)))
	for _, MinesElement := range value.Mines {
		MinesElement.Write(writer)
	}
	mStream.WriteInt32(writer, int32(len(value.LootBoxes)))
	for _, LootBoxesElement := range value.LootBoxes {
		LootBoxesElement.Write(writer)
	}
}

File: clients/cpp/model/Game.cpp
#include "Game.hpp"

Game::Game() { }
Game::Game(int currentTick, Properties properties, Level level, std::vector<Player> players, std::vector<Unit> units, std::vector<Bullet> bullets, std::vector<Mine> mines, std::vector<LootBox> lootBoxes) : currentTick(currentTick), properties(properties), level(level), players(players), units(units), bullets(bullets), mines(mines), lootBoxes(lootBoxes) { }
Game Game::readFrom(InputStream& stream) {
    Game result;
    result.currentTick = stream.readInt();
    result.properties = Properties::readFrom(stream);
    result.level = Level::readFrom(stream);
    result.players = std::vector<Player>(stream.readInt());
    for (size_t i = 0; i < result.players.size(); i++) {
        result.players[i] = Player::readFrom(stream);
    }
    result.units = std::vector<Unit>(stream.readInt());
    for (size_t i = 0; i < result.units.size(); i++) {
        result.units[i] = Unit::readFrom(stream);
    }
    result.bullets = std::vector<Bullet>(stream.readInt());
    for (size_t i = 0; i < result.bullets.size(); i++) {
        result.bullets[i] = Bullet::readFrom(stream);
    }
    result.mines = std::vector<Mine>(stream.readInt());
    for (size_t i = 0; i < result.mines.size(); i++) {
        result.mines[i] = Mine::readFrom(stream);
    }
    result.lootBoxes = std::vector<LootBox>(stream.readInt());
    for (size_t i = 0; i < result.lootBoxes.size(); i++) {
        result.lootBoxes[i] = LootBox::readFrom(stream);
    }
    return result;
}
void Game::writeTo(OutputStream& stream) const {
    stream.write(currentTick);
    properties.writeTo(stream);
    level.writeTo(stream);
    stream.write((int)(players.size()));
    for (const Player& playersElement : players) {
        playersElement.writeTo(stream);
    }
    stream.write((int)(units.size()));
    for (const Unit& unitsElement : units) {
        unitsElement.writeTo(stream);
    }
    stream.write((int)(bullets.size()));
    for (const Bullet& bulletsElement : bullets) {
        bulletsElement.writeTo(stream);
    }
    stream.write((int)(mines.size()));
    for (const Mine& minesElement : mines) {
        minesElement.writeTo(stream);
    }
    stream.write((int)(lootBoxes.size()));
    for (const LootBox& lootBoxesElement : lootBoxes) {
        lootBoxesElement.writeTo(stream);
    }
}
std::string Game::toString() const {
    return std::string("Game") + "(" +
        std::to_string(currentTick) +
        properties.toString() +
        level.toString() +
        "TODO" + 
        "TODO" + 
        "TODO" + 
        "TODO" + 
        "TODO" + 
        ")";
}
